### ISSUE TITLE
Reset gss_krb5_ccache_name value

### ISSUE THREAD

**Issue by jborean93** [heimdal/heimdal#803]
**Is your feature request related to a problem? Please describe.**
Calling `gss_krb5_ccache_name()` will set a thread specific value so that the default ccache used for Kerberos credentials will use that instead of an environment specific value (env var/config file/etc). There is unfortunately no way, that I can find, to unset this value so the default ccache reverts back to the original environment specific value.

MIT allows you to set `name` as `NULL` to unset the thread specific value but Heimdal does nothing when `NULL` is passed in.

**Describe the solution you'd like**
Ideally the behaviour should match MIT GSSAPI where I can do `gss_krb5_ccache_name(&min_status, NULL, NULL)` to unset this value and go back to the environment lookup behaviour. 

**Describe alternatives you've considered**
I tried to do `krb5_cc_set_default_name(ctx, NULL)` with `KRB5CCNAME` set to any the but unfortunately this is set for the `ctx` passed in and I'm unsure of any public way to get the context used in the gssapi APIs.

**Additional context**
It looks like this behaviour has been written but it isn't plugged into anything. I can see [_gsskrb5_krb5_ccache_name](https://github.com/heimdal/heimdal/blob/9dcab76724b417140b4e475701118a01d2892e7c/lib/gssapi/krb5/ccache_name.c) but this method isn't referenced anywhere.

The current logic I can see is as follows:

* `gss_krb5_ccache_name` -> https://github.com/heimdal/heimdal/blob/9dcab76724b417140b4e475701118a01d2892e7c/lib/gssapi/mech/gss_krb5.c#L483-L510
  * This always sets `out_name` to `NULL` which is unlike MIT
  * The method in `krb5/ccache_name.c` does seem to have logic for setting `out_name` like MIT does but this method isn't ever called
* The `gm_set_sec_context_option` glue function is called
  * For Kerberos this is [_gsskrb5_set_sec_context_option](https://github.com/heimdal/heimdal/blob/9dcab76724b417140b4e475701118a01d2892e7c/lib/gssapi/krb5/set_sec_context_option.c#L214-L361)
  * It is invoked with the OID `GSS_KRB5_CCACHE_NAME_X`
* `_gsskrb5_set_sec_context_option` returns `GSS_S_CALL_INACCESSIBLE_READ` when the value is `NULL`
  * https://github.com/heimdal/heimdal/blob/9dcab76724b417140b4e475701118a01d2892e7c/lib/gssapi/krb5/set_sec_context_option.c#L303-L306
  * This return value is lost as `gss_krb5_ccache_name` doesn't use this value at all
  * No changes are made

Even if `_gsskrb5_set_sec_context_option` passed the `NULL` value to [krb5_cc_set_default_name](https://github.com/heimdal/heimdal/blob/9dcab76724b417140b4e475701118a01d2892e7c/lib/krb5/cache.c#L690-L750) it looks like [context->default_cc_name_set](https://github.com/heimdal/heimdal/blob/9dcab76724b417140b4e475701118a01d2892e7c/lib/krb5/cache.c#L713) is only ever reset if both the value is `NULL` and the env var `KRB5CCNAME` is set.

**Comment by lhoward** [comment]
Excellent bug report. `gss_krb5_ccache_name()` is a hot mess that does nothing, it appears.

**Comment by lhoward** [comment]
Untested patch

```diff
From ef1308df894f0b951b749ba46d7a66b38eac2e45 Mon Sep 17 00:00:00 2001
From: Luke Howard <lukeh@padl.com>
Date: Wed, 11 Aug 2021 12:55:02 +1000
Subject: [PATCH] gssapi: implement gss_krb5_ccache_name()

gss_krb5_ccache_name() was a NOOP.

Should close: #803
---
 lib/gssapi/krb5/ccache_name.c            | 27 ++++++++++++------------
 lib/gssapi/krb5/gsskrb5_locl.h           |  5 +++++
 lib/gssapi/krb5/set_sec_context_option.c |  9 ++++++++
 lib/gssapi/mech/gss_krb5.c               | 24 ++++++++++++++-------
 4 files changed, 44 insertions(+), 21 deletions(-)

diff --git a/lib/gssapi/krb5/ccache_name.c b/lib/gssapi/krb5/ccache_name.c
index 1577fb20d..0a2d5960f 100644
--- a/lib/gssapi/krb5/ccache_name.c
+++ b/lib/gssapi/krb5/ccache_name.c
@@ -33,7 +33,7 @@
 
 #include "gsskrb5_locl.h"
 
-char *last_out_name;
+static heim_base_atomic(char *) last_out_name; /* XXX should be thread-specific */
 
 OM_uint32
 _gsskrb5_krb5_ccache_name(OM_uint32 *minor_status,
@@ -48,24 +48,25 @@ _gsskrb5_krb5_ccache_name(OM_uint32 *minor_status,
     GSSAPI_KRB5_INIT(&context);
 
     if (out_name) {
-	const char *n;
+	const char *def_name;
 
-	if (last_out_name) {
-	    free(last_out_name);
-	    last_out_name = NULL;
-	}
+	*out_name = NULL;
 
-	n = krb5_cc_default_name(context);
-	if (n == NULL) {
-	    *minor_status = ENOMEM;
-	    return GSS_S_FAILURE;
+	def_name = krb5_cc_default_name(context);
+	if (def_name) {
+	    char *s = strdup(def_name);
+	    if (s) {
+		s = heim_base_exchange_pointer(&last_out_name, s);
+		free(s);
+
+		*out_name = last_out_name;
+	    }
 	}
-	last_out_name = strdup(n);
-	if (last_out_name == NULL) {
+
+	if (*out_name == NULL) {
 	    *minor_status = ENOMEM;
 	    return GSS_S_FAILURE;
 	}
-	*out_name = last_out_name;
     }
 
     kret = krb5_cc_set_default_name(context, name);
diff --git a/lib/gssapi/krb5/gsskrb5_locl.h b/lib/gssapi/krb5/gsskrb5_locl.h
index b2e7c178b..fbbb168c6 100644
--- a/lib/gssapi/krb5/gsskrb5_locl.h
+++ b/lib/gssapi/krb5/gsskrb5_locl.h
@@ -145,4 +145,9 @@ extern HEIMDAL_MUTEX gssapi_keytab_mutex;
 #define SC_ORDER          0x0080
 #define SC_AUTHENTICATOR  0x0100
 
+struct gsskrb5_ccache_name_args {
+    const char *name;
+    const char *out_name;
+};
+
 #endif
diff --git a/lib/gssapi/krb5/set_sec_context_option.c b/lib/gssapi/krb5/set_sec_context_option.c
index 56a30cd07..e0531724f 100644
--- a/lib/gssapi/krb5/set_sec_context_option.c
+++ b/lib/gssapi/krb5/set_sec_context_option.c
@@ -352,6 +352,15 @@ _gsskrb5_set_sec_context_option
 
 	*minor_status = 0;
 	return GSS_S_COMPLETE;
+    } else if (gss_oid_equal(desired_object, GSS_KRB5_CCACHE_NAME_X)) {
+	struct gsskrb5_ccache_name_args *args = value->value;
+
+	if (value->length != sizeof(*args)) {
+	    *minor_status = EINVAL;
+	    return GSS_S_FAILURE;
+	}
+
+	return _gsskrb5_krb5_ccache_name(minor_status, args->name, &args->out_name);
     } else if (gss_oid_equal(desired_object, GSS_KRB5_IMPORT_RFC4121_CONTEXT_X)) {
 	return make_rfc4121_context(minor_status, context, context_handle, value);
     }
diff --git a/lib/gssapi/mech/gss_krb5.c b/lib/gssapi/mech/gss_krb5.c
index ee376229c..36fa85c92 100644
--- a/lib/gssapi/mech/gss_krb5.c
+++ b/lib/gssapi/mech/gss_krb5.c
@@ -487,26 +487,34 @@ gss_krb5_ccache_name(OM_uint32 *minor_status,
 {
     struct _gss_mech_switch *m;
     gss_buffer_desc buffer = GSS_C_EMPTY_BUFFER;
-    OM_uint32 junk;
+    OM_uint32 major_status;
+    struct gsskrb5_ccache_name_args args;
 
     _gss_load_mech();
 
+    *minor_status = 0;
+
     if (out_name)
 	*out_name = NULL;
 
-    buffer.value = rk_UNCONST(name);
-    if (name) {
-	buffer.length = strlen(name);
-    }
+    args.name = name;
+    args.out_name = NULL;
+
+    buffer.value = &args;
+    buffer.length = sizeof(args);
+
+    major_status = GSS_S_UNAVAILABLE;
 
     HEIM_TAILQ_FOREACH(m, &_gss_mechs, gm_link) {
 	if (m->gm_mech.gm_set_sec_context_option == NULL)
 	    continue;
-	m->gm_mech.gm_set_sec_context_option(&junk, NULL,
-	    GSS_KRB5_CCACHE_NAME_X, &buffer);
+	major_status = m->gm_mech.gm_set_sec_context_option(minor_status,
+	    NULL, GSS_KRB5_CCACHE_NAME_X, &buffer);
     }
 
-    return (GSS_S_COMPLETE);
+    *out_name = args.out_name;
+
+    return major_status;
 }
 
 
-- 
2.30.1 (Apple Git-130)

```

**Comment by jborean93** [comment]
The first problem is `major_status` is being set to the value of the last mech in the loop that defined `gm_set_sec_context_option`.

```c
HEIM_TAILQ_FOREACH(m, &_gss_mechs, gm_link) {
    if (m->gm_mech.gm_set_sec_context_option == NULL)
        continue;
    major_status = m->gm_mech.gm_set_sec_context_option(minor_status,
        NULL, GSS_KRB5_CCACHE_NAME_X, &buffer);
    }
}
```

In my case that's NTLM which is returning `GSS_S_UNAVAILABLE` as the context is `NULL`. Even if the context was set it would still return that value as the OID isn't valid for NTLM. The same problem applies to the SPNEGO mech but it returns `GSS_S_NO_CONTEXT`. I'm not sure what the proper behaviour should be, should it:

* Only return the `major_status` if no mech returns `0`
  * This should work if krb5 doesn't report an error
* Ignore `GSS_S_NO_CONTEXT` and `GSS_S_UNAVAILABLE` and pretend they are the not implemented like mechs without `gm_set_sec_context_option`
* Is there some other mechanism in Heimdal to filter out non Kerberos mechs and just run it on that

It also begs the question should a non-0 `major_status` return immediately or is returning the last status value the correct behaviour?

After changing that code locally to consider any successful call means return 0 I came across another problem. It is still calling the older code and it's setting the ccache to the value of the `gsskrb5_ccache_name_args` struct that you've now added. When removing that older code I can confirm the behaviour is more like MIT krb5 and the old name is being returned as expected.

Here is a diff of the changes I've made based on your diff

```diff
diff --git a/lib/gssapi/krb5/ccache_name.c b/lib/gssapi/krb5/ccache_name.c
index 1577fb20d..0a2d5960f 100644
--- a/lib/gssapi/krb5/ccache_name.c
+++ b/lib/gssapi/krb5/ccache_name.c
@@ -33,7 +33,7 @@
 
 #include "gsskrb5_locl.h"
 
-char *last_out_name;
+static heim_base_atomic(char *) last_out_name; /* XXX should be thread-specific */
 
 OM_uint32
 _gsskrb5_krb5_ccache_name(OM_uint32 *minor_status,
@@ -48,24 +48,25 @@ _gsskrb5_krb5_ccache_name(OM_uint32 *minor_status,
     GSSAPI_KRB5_INIT(&context);
 
     if (out_name) {
-	const char *n;
+	const char *def_name;
 
-	if (last_out_name) {
-	    free(last_out_name);
-	    last_out_name = NULL;
-	}
+	*out_name = NULL;
 
-	n = krb5_cc_default_name(context);
-	if (n == NULL) {
-	    *minor_status = ENOMEM;
-	    return GSS_S_FAILURE;
+	def_name = krb5_cc_default_name(context);
+	if (def_name) {
+	    char *s = strdup(def_name);
+	    if (s) {
+		s = heim_base_exchange_pointer(&last_out_name, s);
+		free(s);
+
+		*out_name = last_out_name;
+	    }
 	}
-	last_out_name = strdup(n);
-	if (last_out_name == NULL) {
+
+	if (*out_name == NULL) {
 	    *minor_status = ENOMEM;
 	    return GSS_S_FAILURE;
 	}
-	*out_name = last_out_name;
     }
 
     kret = krb5_cc_set_default_name(context, name);
diff --git a/lib/gssapi/krb5/gsskrb5_locl.h b/lib/gssapi/krb5/gsskrb5_locl.h
index b2e7c178b..fbbb168c6 100644
--- a/lib/gssapi/krb5/gsskrb5_locl.h
+++ b/lib/gssapi/krb5/gsskrb5_locl.h
@@ -145,4 +145,9 @@ extern HEIMDAL_MUTEX gssapi_keytab_mutex;
 #define SC_ORDER          0x0080
 #define SC_AUTHENTICATOR  0x0100
 
+struct gsskrb5_ccache_name_args {
+    const char *name;
+    const char *out_name;
+};
+
 #endif
diff --git a/lib/gssapi/krb5/set_sec_context_option.c b/lib/gssapi/krb5/set_sec_context_option.c
index 56a30cd07..3a6f86bb2 100644
--- a/lib/gssapi/krb5/set_sec_context_option.c
+++ b/lib/gssapi/krb5/set_sec_context_option.c
@@ -294,23 +294,6 @@ _gsskrb5_set_sec_context_option
 	*minor_status = EINVAL;
 	return GSS_S_FAILURE;
 
-    } else if (gss_oid_equal(desired_object, GSS_KRB5_CCACHE_NAME_X)) {
-	char *str;
-
-	maj_stat = get_string(minor_status, value, &str);
-	if (maj_stat != GSS_S_COMPLETE)
-	    return maj_stat;
-	if (str == NULL) {
-	    *minor_status = 0;
-	    return GSS_S_CALL_INACCESSIBLE_READ;
-	}
-
-	*minor_status = krb5_cc_set_default_name(context, str);
-	free(str);
-	if (*minor_status)
-	    return GSS_S_FAILURE;
-
-	return GSS_S_COMPLETE;
     } else if (gss_oid_equal(desired_object, GSS_KRB5_SET_TIME_OFFSET_X)) {
 	OM_uint32 offset;
 	time_t t;
@@ -352,6 +335,15 @@ _gsskrb5_set_sec_context_option
 
 	*minor_status = 0;
 	return GSS_S_COMPLETE;
+    } else if (gss_oid_equal(desired_object, GSS_KRB5_CCACHE_NAME_X)) {
+	struct gsskrb5_ccache_name_args *args = value->value;
+
+	if (value->length != sizeof(*args)) {
+	    *minor_status = EINVAL;
+	    return GSS_S_FAILURE;
+	}
+
+	return _gsskrb5_krb5_ccache_name(minor_status, args->name, &args->out_name);
     } else if (gss_oid_equal(desired_object, GSS_KRB5_IMPORT_RFC4121_CONTEXT_X)) {
 	return make_rfc4121_context(minor_status, context, context_handle, value);
     }
diff --git a/lib/gssapi/mech/gss_krb5.c b/lib/gssapi/mech/gss_krb5.c
index ee376229c..fcc917a53 100644
--- a/lib/gssapi/mech/gss_krb5.c
+++ b/lib/gssapi/mech/gss_krb5.c
@@ -487,26 +487,37 @@ gss_krb5_ccache_name(OM_uint32 *minor_status,
 {
     struct _gss_mech_switch *m;
     gss_buffer_desc buffer = GSS_C_EMPTY_BUFFER;
-    OM_uint32 junk;
+    OM_uint32 major_status;
+    OM_uint32 mech_major_status;
+    struct gsskrb5_ccache_name_args args;
 
     _gss_load_mech();
 
+    *minor_status = 0;
+
     if (out_name)
 	*out_name = NULL;
 
-    buffer.value = rk_UNCONST(name);
-    if (name) {
-	buffer.length = strlen(name);
-    }
+    args.name = name;
+    args.out_name = NULL;
+
+    buffer.value = &args;
+    buffer.length = sizeof(args);
+
+    major_status = GSS_S_UNAVAILABLE;
 
     HEIM_TAILQ_FOREACH(m, &_gss_mechs, gm_link) {
 	if (m->gm_mech.gm_set_sec_context_option == NULL)
 	    continue;
-	m->gm_mech.gm_set_sec_context_option(&junk, NULL,
-	    GSS_KRB5_CCACHE_NAME_X, &buffer);
+	mech_major_status = m->gm_mech.gm_set_sec_context_option(minor_status,
+	    NULL, GSS_KRB5_CCACHE_NAME_X, &buffer);
+	if (major_status != GSS_S_COMPLETE)
+	    major_status = mech_major_status;
     }
 
-    return (GSS_S_COMPLETE);
+    *out_name = args.out_name;
+
+    return major_status;
 }
 
 

```

The only difference is that it treats at least 1 `GSS_S_COMPLETE` from the `set_sec_context_option` as a overall success and removed the older `GSS_KRB5_CCACHE_NAME_X` code.

**Comment by lhoward** [comment]
I'm not sure why the code in gss_krb5.c doesn't just look for the Kerberos mechanism specifically but, anyway, your patch looks good I think.

**Comment by lhoward** [comment]
Might be better to check for `mech_major_status != GSS_S_UNAVAILABLE`.

**Comment by lhoward** [comment]
Fixed in 932cbb2630332cc3a7df4faa73c7808697f3860d, merged, please test.

**Comment by jborean93** [comment]
Thanks @lhoward. Apologies for not following up earlier but I just had 1 tiny concern around checking only for `mech_major_status != GSS_S_UNAVAILABLE`. During my original testing I found the the SPNEGO mech would return `GSS_S_NO_CONTEXT` because the context was `NULL` https://github.com/heimdal/heimdal/blob/7f8bb85cf53b85a8b33143eb99c110bca96de254/lib/gssapi/spnego/context_stubs.c#L529-L531. The code that has been merged works just fine because the krb5 mech is before the SPNEGO one during the loop but I'm unsure whether this is something that will continue to be true in the future. Is this something that the function should check on the return or do you think it's fine to rely on the foreach ordering of the mechs?
